src/typechecker.erl

```diff
@@ -3456,8 +3456,21 @@ refinable(?type(union,Tys), TEnv, Trace) when is_list(Tys) ->
     lists:all(fun (Ty) -> refinable(Ty, TEnv, Trace) end, Tys);
 refinable(?type(tuple, Tys), TEnv, Trace) when is_list(Tys) ->
     lists:all(fun (Ty) -> refinable(Ty, TEnv, Trace) end, Tys);
-refinable(?type(record, [_ | Fields]), TEnv, Trace) ->
-    lists:all(fun (Ty) -> refinable(Ty, TEnv, Trace) end, [X || ?type(field_type, X) <- Fields]);
+refinable(?type(record, [{atom, _, RecordName}]), TEnv = #{records := REnv}, Trace) ->
```

Comment:
Why not allow any refined Fields here, e.g. types like `#rec{field :: a}`?

Reply:
From my testing (with Gradualizer's unit tests and a significant code base), `refinable` is never called with the full type for a record i.e. the type does not have any parameters beyond the record name.
`begin ?type(record, [_ | X]), length(X) =:= 0 end` is always true.

Reply:
That seems weird. I would guess it should call `refinable` with any type, but I haven't looked at the details.